File: 213. House Robber II/House Robber II.js
function rob(nums) {
  if (nums.length === 1) {
    return nums[0];
  }

  let rob1 = 0;
  let rob2 = 0;
  for (let value of nums) {
    const temp = Math.max(rob1 + value, rob2);
    rob1 = rob2;
    rob2 = temp;
  }

  let robNoFirst = 0;
  let robNoLast = 0;
  for (let i = 1; i < nums.length; i++) {
    const temp = Math.max(robNoFirst + nums[i], robNoLast);
    robNoFirst = robNoLast;
    robNoLast = temp;
  }

  return Math.max(rob1, robNoLast);
}

File: 213. House Robber II/House Robber II.java
import java.util.Arrays;

class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }

        return Math.max(calculate(Arrays.copyOfRange(nums, 1, nums.length)), calculate(Arrays.copyOfRange(nums, 0, nums.length - 1)));
    }

    private int calculate(int[] nums) {
        int rob1 = 0;
        int rob2 = 0;
        for (int value : nums) {
            int temp = Math.max(rob1 + value, rob2);
            rob1 = rob2;
            rob2 = temp;
        }
        return rob2;
    }
}
